workflow/util/util.go

```diff
@@ -410,32 +428,71 @@ func updateWorkflowNodeByKey(wfIf v1alpha1.WorkflowInterface, hydrator hydrator.
 		for nodeID, node := range wf.Status.Nodes {
 			if node.IsActiveSuspendNode() {
 				if SelectorMatchesNode(selector, node) {
-					node.Phase = phase
-					node.FinishedAt = metav1.Time{Time: time.Now().UTC()}
-					if len(message) > 0 {
-						node.Message = message
+
+					// Update phase
+					if values.Phase != "" {
+						node.Phase = values.Phase
+						if values.Phase.Fulfilled() {
+							node.FinishedAt = metav1.Time{Time: time.Now().UTC()}
+						}
+						nodeUpdated = true
+					}
+
+					// Update message
+					if values.Message != "" {
+						node.Message = values.Message
+						nodeUpdated = true
 					}
+
+					// Update output parameters
+					if len(values.OutputParameters) > 0 {
+						if node.Outputs == nil {
+							return true, fmt.Errorf("cannot set output parameters because node is not expecting any raw parameters")
+						}
+						for name, val := range values.OutputParameters {
+							hit := false
+							for i, param := range node.Outputs.Parameters {
+								if param.Name == name {
+									if param.ValueFrom == nil || param.ValueFrom.Supplied == nil {
+										return true, fmt.Errorf("cannot set output parameter '%s' because it does not use valueFrom.raw or it was already set", param.Name)
+									}
+									intStr := intstr.FromString(val)
+									node.Outputs.Parameters[i].Value = &intStr
+									node.Outputs.Parameters[i].ValueFrom = nil
+									nodeUpdated = true
+									hit = true
+									break
+								}
+							}
+							if !hit {
+								return true, fmt.Errorf("node is not expecting output parameter '%s'", name)
+							}
+						}
+					}
+
 					wf.Status.Nodes[nodeID] = node
-					nodeUpdated = true
 				}
 			}
 		}
-		if nodeUpdated {
-			err := hydrator.Dehydrate(wf)
-			if err != nil {
-				return false, fmt.Errorf("unable to compress or offload workflow nodes: %s", err)
-			}
 
-			_, err = wfIf.Update(wf)
-			if err != nil {
-				if apierr.IsConflict(err) {
-					return false, nil
-				}
-				return false, err
+		if !nodeUpdated {
+			return true, fmt.Errorf("currently, set only targets suspend nodes: no suspend nodes matching nodeFieldSelector: %s", nodeFieldSelector)
+		}
+
+		err = hydrator.Dehydrate(wf)
+		if err != nil {
+			return true, fmt.Errorf("unable to compress or offload workflow nodes: %s", err)
+		}
+
+		_, err = wfIf.Update(wf)
```

Comment:
Can user set the value while workflow running?

Reply:
Yes, if a Workflow is "suspended" then it's running, which is the target phase for this feature. You bring up a good point though... we should retry in case the Workflow object changes while we're in this function. Will refactor.